### PR TITLE
Barber

### PR DESCRIPTION
**Pull request by AstralSkies**

### CODE REVIEW

File: EOLib.Localization/EOResourceID.cs [hunk 1]
```diff
@@ -152,8 +152,13 @@ public enum EOResourceID
...
         DIALOG_SHOP_CRAFT_PUT_INGREDIENTS_TOGETHER = 159,
 
+        DIALOG_BARBER_CHANGE_MODAL = 160,
+        DIALOG_BARBER_CHANGE_HAIR_COLOR = 161,
+        DIALOG_BARBER_BUY_HAIRSTYLE = 162,
         DIALOG_WORD_CURRENT = 163,
-
+        DIALOG_BARBER_DO_YOU_WANT_TO_BUY_A_NEW_HAIRSTYLE = 164,
+        
+        DIALOG_TRADE_PLAYERS_NEED_TO_TRADE_ONE_ITEM = 165,
```
Comment (ethanmoffat) [inline EOLib.Localization/EOResourceID.cs:161]: This new line is a duplicate of the resource below it and doesn't appear to be used in this PR? I think this line can be deleted

File: EOLib/Domain/Interact/Bank/BankDataRepository.cs [hunk 2]
```diff
@@ -1,34 +1,47 @@
-﻿using AutomaticTypeMapper;
+﻿using System.Diagnostics;
+using AutomaticTypeMapper;
```
Comment (ethanmoffat) [inline EOLib/Domain/Interact/Bank/BankDataRepository.cs:2]: Changes in this file can be reverted in their entirety

File: EOLib.Localization/EOResourceID.cs [hunk 3]
```diff
@@ -342,6 +347,8 @@ public enum EOResourceID
         STATUS_LABEL_IS_ONLINE_NOT_FOUND = 430,
         DIALOG_BANK_LOCKER_UPGRADE = 431,
         DIALOG_BANK_MORE_SPACE = 432,
-        STATUS_LABEL_LOCKER_SPACE_INCREASED = 433
+        STATUS_LABEL_LOCKER_SPACE_INCREASED = 433,
+
```
Comment (ethanmoffat) [inline EOLib.Localization/EOResourceID.cs:351]: nit: delete extra whitespace lines

File: EOLib/Domain/Interact/Barber/BarberDataRepository.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    [AutoMappedType(IsSingleton = true)]
+    public class BarberDataRepository : IBarberDataRepository, IBarberDataProvider
+    {
+        private int _sessionID;
+        public int SessionID
+        {
+            get => _sessionID;
+            set
+            {
+                _sessionID = value;
```
Comment (ethanmoffat) [inline EOLib/Domain/Interact/Barber/BarberDataRepository.cs:25]: Make this an auto-property:

`public int SessionID { get; set; }`

File: EOLib/PacketHandlers/Avatar/AvatarAgreeHandler.cs [hunk 5]
```diff
@@ -17,7 +17,7 @@ public class AvatarAgreeHandler : InGameOnlyPacketHandler
         protected readonly ICurrentMapStateRepository _currentMapStateRepository;
         protected readonly ICharacterRepository _characterRepository;
         protected readonly IEIFFileProvider _eifFileProvider;
-
+
```
Comment (ethanmoffat) [inline EOLib/PacketHandlers/Avatar/AvatarAgreeHandler.cs:20]: Changes in this file can be reverted in their entirety

File: EOLib/PacketHandlers/Barber/BarberReplyHandler.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+using EOLib.Domain.Interact.Barber;
+using EOLib.Domain.Character;
+using EOLib.Domain.Login;
+using EOLib.Domain.Map;
+using EOLib.Domain.Notifiers;
+
+namespace EOLib.PacketHandlers.Barber
+{
+    [AutoMappedType]
+    public class BarberReplyHandler : InGameOnlyPacketHandler
```
Comment (ethanmoffat) [inline EOLib/PacketHandlers/Barber/BarberReplyHandler.cs:15]: Rename type/file to `BarberAgreeHandler` to match the packet ID

File: EOLib/PacketHandlers/Shop/ShopTradeHandler.cs [hunk 7]
```diff
@@ -109,4 +109,4 @@ public class ShopSellHandler : ShopTradeHandler
         {
         }
```
Comment (ethanmoffat) [inline EOLib/PacketHandlers/Shop/ShopTradeHandler.cs:110]: Changes in this file can be reverted in their entirety

File: EndlessClient.sln [hunk 8]
```diff
@@ -13,7 +13,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "EOLib", "EOLib\EOLib.csproj
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "EOLib.Config", "EOLib.Config\EOLib.Config.csproj", "{B6AD3196-D307-4352-9979-9237CB08B102}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "EOLib.Graphics", "EOLib.Graphics\EOLib.Graphics.csproj", "{8C4BC247-89ED-4629-B34B-15A5AEE88E2D}"
```
Comment (ethanmoffat) [inline EndlessClient.sln:16]: Not sure what's happening here, can these changes be reverted?

File: EndlessClient/ControlSets/InitialControlSet.cs [hunk 9]
```diff
@@ -159,6 +159,8 @@ private IXNALabel GetVersionInfoLabel()
         private IXNAPictureBox GetPersonPicture1()
         {
             var texture = _personSet1[_randomGen.Next(4)];
+
```
Comment (ethanmoffat) [inline EndlessClient/ControlSets/InitialControlSet.cs:162]: Changes to this file can be reverted in their entirety

File: EndlessClient/Dialogs/Actions/InGameDialogActions.cs [hunk 10]
```diff
@@ -12,6 +12,7 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using System.Diagnostics;
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/Actions/InGameDialogActions.cs:15]: nit: remove unused using

File: EndlessClient/Dialogs/Actions/InGameDialogActions.cs [hunk 11]
```diff
@@ -69,7 +71,9 @@ public class InGameDialogActions : IInGameDialogActions
                                    ILawDialogFactory lawDialogFactory,
                                    IHelpDialogFactory helpDialogFactory,
                                    ISfxPlayer sfxPlayer,
-                                   IStatusLabelSetter statusLabelSetter)
+                                   IStatusLabelSetter statusLabelSetter,
+                                   IBarberDialogFactory barberDialogFactory)
+
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/Actions/InGameDialogActions.cs:76]: nit: remove whitespace

File: EndlessClient/Dialogs/Actions/NpcInteractionActions.cs [hunk 12]
```diff
@@ -11,6 +11,7 @@
 using EOLib.Localization;
 using System;
 using XNAControls;
+using System.Diagnostics;
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/Actions/NpcInteractionActions.cs:14]: nit: remove unused using

File: EndlessClient/Dialogs/ActiveDialogRepository.cs [hunk 13]
```diff
@@ -165,6 +171,8 @@ IReadOnlyList<Option<IXNADialog>> ActiveDialogs
                     JukeboxDialog.Map(Map),
                     InnkeeperDialog.Map(Map),
                     LawDialog.Map(Map),
+                    BarberDialog.Map(Map),
+                    BardDialog.Map(Map),
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/ActiveDialogRepository.cs:175]: `BardDialog.Map` is already in this list

File: EndlessClient/Dialogs/BaseEODialog.cs [hunk 14]
```diff
@@ -12,6 +12,8 @@ public abstract class BaseEODialog : XNADialog
 
         public INativeGraphicsManager GraphicsManager { get; }
 
+        public ListDialogItem.ListItemStyle ListItemType { get; set; }
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/BaseEODialog.cs:15]: This shouldn't be in `BaseEODialog`. Not every dialog has list items.

File: EndlessClient/Rendering/Character/CharacterAnimator.cs [hunk 15]
```diff
@@ -17,6 +17,7 @@
 using System.Diagnostics;
 using System.Linq;
 
+
```
Comment (ethanmoffat) [inline EndlessClient/Rendering/Character/CharacterAnimator.cs:20]: Changes to this file can be removed in their entirety

File: EndlessClient/Rendering/Sprites/SpriteSheet.cs [hunk 16]
```diff
@@ -1,5 +1,6 @@
 ﻿using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
+using System.Diagnostics;
```
Comment (ethanmoffat) [inline EndlessClient/Rendering/Sprites/SpriteSheet.cs:3]: Changes to this file can be removed in their entirety

File: EndlessClient/UIControls/CreateCharacterControl.cs [hunk 17]
```diff
@@ -72,5 +72,22 @@ private static CharacterRenderProperties GetDefaultProperties()
         {
             return new CharacterRenderProperties.Builder { HairStyle = 1 }.ToImmutable();
         }
+
+        public void UpdateRenderProperties(int hairStyle, int hairColor, int race, int gender, int armorgraphic, int weapongraphic, int bootgraphic, int shieldgraphic)
```
Comment (ethanmoffat) [inline EndlessClient/UIControls/CreateCharacterControl.cs:76]: Instead of adding an `UpdateRenderProperties` method, add a constructor overload (and factory function overload) that takes in the current character's render properties instead of using default properties. Then you can just construct the object directly in `BarberDialog` from the main character's render properties.

It looks like barber dialog doesn't need to arbitrarily set the render properties, and I'd prefer not to have a mutating method like this on the control's interface.

For clarify, I'm thinking like this:
```csharp
public CreateCharacterControl(ICharacterRendererFactory characterRendererFactory)
    : this(GetDefaultProperties(), characterRendererFactory) { }

public CreateCharacterControl(CharacterRenderProperties renderProperties, ICharacterRendererFactory factory)
    : base(Character.Default.WithRenderProperties(renderProperties), factory
{
    SetSize(99, 123);
    _lastPosition = Vector2.Zero;
}
```

File: EndlessClient/Dialogs/ListDialogItem.cs [hunk 18]
```diff
@@ -18,6 +18,10 @@ public enum ListItemStyle
             Large
         }
 
+
+        public bool EnableBarberMode = false;
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/ListDialogItem.cs:22]: `ListDialogItem` should not have anything barber-specific. This is a general control for all list dialog items. Also, convention is to avoid public fields and use properties instead.

If you want to override the width of the highlight, add a property that sets it, similar to `OffsetX` and `OffsetY`.

`OffsetX` already performs the function you're doing in the `Draw` override for getting `rectX`.

File: EndlessClient/UIControls/BlinkingLabel.cs [hunk 19]
```diff
@@ -1,4 +1,5 @@
 ﻿using System;
+using System.Diagnostics;
```
Comment (ethanmoffat) [inline EndlessClient/UIControls/BlinkingLabel.cs:2]: I don't see the value in these changes - is there something wrong with `BlinkingLabel`? Strongly prefer to revert this file in its entirety.

Usually a crash in `base.Update()` shows up in `EndlessGame` due to collections being modified. This is *usually* caused by trying to update the collection on a background thread (collection was modified, enumeration may not continue).

File: EndlessClient/Dialogs/BarberDialog.cs [hunk 20]
```diff
@@ -0,0 +1,223 @@
...
+        }
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            _characterControl.Initialize();
+            UpdateListItems(_characterRepository.MainCharacter.RenderProperties);
+
+        }
+        public override void Update(GameTime gameTime)
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/BarberDialog.cs:94]: This override is unneeded

File: EndlessClient/Dialogs/BarberDialog.cs [hunk 21]
```diff
@@ -0,0 +1,223 @@
...
+            var message = $"{_localizedStringFinder.GetString(EOResourceID.DIALOG_BARBER_DO_YOU_WANT_TO_BUY_A_NEW_HAIRSTYLE)}, {cost} {_eifFileProvider.EIFFile[1].Name}";
+            var title = _localizedStringFinder.GetString(EOResourceID.DIALOG_BARBER_BUY_HAIRSTYLE);
+            var msgBox = _messageBoxFactory.CreateMessageBox(message, title, EODialogButtons.OkCancel);
+
+            msgBox.DialogClosing += (_, e) =>
+            {
+                if (e.Result == XNADialogResult.OK)
+                {
+                    _barberActions.Purchase(hairStyle, hairColor);
+                    foreach (var notifier in _soundNotifiers)
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/BarberDialog.cs:208]: Notifiers are generally used for signalling from `EOLib` up to `EndlessClient` that something happened (because I wanted to re-invent event handlers, I guess, I regret the design decision but I digress)

You can just inject `ISfxPlayer` and play `SoundEffectID.BuySell`. See jukebox dialog.

File: EndlessClient/Dialogs/Factories/HelpDialogFactory.cs [hunk 22]
```diff
@@ -53,13 +53,11 @@ private string[] GetMessages()
         {
             return new[]
             {
-                _localizedStringFinder.GetString(EOResourceID.ENDLESS_HELP_SUMMARY_1),
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/Factories/HelpDialogFactory.cs:56]: Why were these changes deleted? Can this file be reverted?

File: EndlessClient/Dialogs/ScrollingListDialog.cs [hunk 23]
```diff
@@ -74,7 +74,7 @@ public enum DialogType
         BankAccountDialog,
 
         Jukebox,
-        Barber,
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/ScrollingListDialog.cs:77]: If `Barber` isn't going to be a `ScrollingListDialog` then this is fine, remove all references to `Barber` from this file

File: EndlessClient/Dialogs/ScrollingListDialog.cs [hunk 24]
```diff
@@ -121,7 +121,7 @@ public int ItemsToShow
             }
         }
 
-        public ListDialogItem.ListItemStyle ListItemType
+        public new ListDialogItem.ListItemStyle ListItemType
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/ScrollingListDialog.cs:124]: This should be reverted along with removing `ListItemType` from `BaseEoDialog`

File: EndlessClient/Dialogs/BarberDialog.cs [hunk 25]
```diff
@@ -0,0 +1,223 @@
...
+        }
+
+        private void Hair(object sender, MonoGame.Extended.Input.InputListeners.MouseEventArgs e) => UpdateHairStyle(1);
+        private void HairColor(object sender, MonoGame.Extended.Input.InputListeners.MouseEventArgs e) => UpdateHairColor(1);
+
+        private void UpdateHairStyle(int offset)
+        {
+            var currentProperties = _characterControl.RenderProperties;
+            var newHairStyle = (currentProperties.HairStyle + offset) % 21;
+            _characterControl.UpdateRenderProperties(newHairStyle, currentProperties.HairColor,
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/BarberDialog.cs:160]: `CreateCharacterControl.NextHairStyle` already does this.

Note that the base class `CharacterControl` has a publicly-visible `RenderProperties` property that you can access to get the current hairstyle for setting the subtext

File: EndlessClient/Dialogs/BarberDialog.cs [hunk 26]
```diff
@@ -0,0 +1,223 @@
...
+                                                     currentProperties.ArmorGraphic, currentProperties.WeaponGraphic,
+                                                     currentProperties.BootsGraphic, currentProperties.ShieldGraphic);
+            _changeHairItem.SubText = GetCurrentHairStyleText(newHairStyle);
+        }
+
+        private void UpdateHairColor(int offset)
+        {
+            var currentProperties = _characterControl.RenderProperties;
+            var newHairColor = (currentProperties.HairColor + offset) % _hairColorNames.Length;
+            _characterControl.UpdateRenderProperties(currentProperties.HairStyle, newHairColor,
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/BarberDialog.cs:171]: `CreateCharacterControl.NextHairColor` already does this.

Note that the base class `CharacterControl` has a publicly-visible `RenderProperties` property that you can access to get the current hair color for setting the subtext

**Review by ethanmoffat — CHANGES_REQUESTED** [review]

File: EndlessClient/Dialogs/BarberDialog.cs [hunk 27]
```diff
@@ -0,0 +1,223 @@
...
+            _characterControl.SetParentControl(this);
+            _characterControl.UpdateRenderProperties(currentProperties.HairStyle, currentProperties.HairColor,
+                                                     currentProperties.Race, currentProperties.Gender,
+                                                     currentProperties.ArmorGraphic, currentProperties.WeaponGraphic,
+                                                     currentProperties.BootsGraphic, currentProperties.ShieldGraphic);
+        }
+
+        private void InitializeDialogItems(IEODialogButtonService dialogButtonService)
+        {
+            var cancel = CreateButton(dialogButtonService, new Vector2(215, 150), SmallButton.Cancel);
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/BarberDialog.cs:81]: I think y coordinate needs to be +1 here
image

File: EndlessClient/UIControls/BlinkingLabel.cs (hunk 19 above)
Comment (AstralSkies) [reply EndlessClient/UIControls/BlinkingLabel.cs:2]: image

I get this issue if i revert or I can do the below instead

image

File: EndlessClient/Dialogs/Factories/HelpDialogFactory.cs (hunk 22 above)
Comment (AstralSkies) [reply EndlessClient/Dialogs/Factories/HelpDialogFactory.cs:56]: leftovers on getting used to the repo haha. sorry !

File: EndlessClient/Dialogs/ListDialogItem.cs [hunk 28]
```diff
@@ -69,6 +66,13 @@ public int OffsetY
             }
         }
 
+        public int HighlightWidthOverride
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/ListDialogItem.cs:69]: 1. Can this be an auto property? Trying to avoid the property-with-backing-field paradigm unless the getter/setter have logic associated with the property and it is unavoidable.
2. Can you make this an `Optional<int>` or `int?` instead of using the magic value `-1` to indicate that it is unset?

File: EndlessClient/Dialogs/BarberDialog.cs [hunk 29]
```diff
@@ -0,0 +1,217 @@
...
+        {
+            _characterControl.NextHairStyle();
+            _changeHairItem.SubText = GetCurrentHairStyleText(_characterControl.RenderProperties.HairStyle);
+        }
+
+        private void HairColor(object sender, MonoGame.Extended.Input.InputListeners.MouseEventArgs e)
+        {
+            UpdateHairColor();
+        }
+        private void UpdateHighlightWidth(ListDialogItem item)
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/BarberDialog.cs:163]: It looks like this method is unused

File: EndlessClient/Dialogs/BarberDialog.cs [hunk 30]
```diff
@@ -0,0 +1,217 @@
...
+        {
+            item.HighlightWidthOverride = 175;
+        }
+        private void UpdateHairColor()
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/BarberDialog.cs:167]: You can inline this method, it is only called once

File: EndlessClient/Dialogs/BarberDialog.cs [hunk 31]
```diff
@@ -0,0 +1,217 @@
...
+        {
+            var currentProperties = _characterControl.RenderProperties;
+            _characterControl.NextHairColor();
+            _changeHairColor.SubText = GetCurrentHairColorText(_characterControl.RenderProperties.HairColor);
+        }
+
+        private void BuyHairStyleOrColor(object sender, MonoGame.Extended.Input.InputListeners.MouseEventArgs e) => PurchaseHairStyleOrColor(_characterControl.RenderProperties.HairStyle, _characterControl.RenderProperties.HairColor);
+
+        private void PurchaseHairStyleOrColor(int hairStyle, int hairColor)
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/BarberDialog.cs:176]: Also prefer to inline this method. I don't think having the event handles hook into a different method really adds much clarity.

File: EndlessClient/Dialogs/BarberDialog.cs [hunk 32]
```diff
@@ -0,0 +1,217 @@
...
+        private string GetCurrentHairColorText(int hairColor) => $"{_localizedStringFinder.GetString(EOResourceID.DIALOG_WORD_CURRENT)}: {_hairColorNames[hairColor % _hairColorNames.Length]}";
+        private string GetCostText() => $"{_localizedStringFinder.GetString(EOResourceID.DIALOG_WORD_CURRENT)}: {CalculateCost()} gold";
+
+        private int CalculateCost()
+        {
+            var level = (int)_characterRepository.MainCharacter.Stats[CharacterStat.Level];
+            return 200 + Math.Max(level - 1, 0) * 200;
+        }
+
+        private void Hair(object sender, MonoGame.Extended.Input.InputListeners.MouseEventArgs e)
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/BarberDialog.cs:153]: Rename the event handlers something like `ChangeHairStyle_Click`, `ChangeHairColor_Click`, `BuyHair_Click` so it's clearer from the names that they're vent handlers.

File: EndlessClient/Dialogs/ListDialogItem.cs [hunk 33]
```diff
@@ -20,6 +21,8 @@ public enum ListItemStyle
 
         private int _index;
         private int _xOffset, _yOffset;
+        private int _highlightWidthOverride = -1;
+        private int _highlightXOffset = 0;
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/ListDialogItem.cs:25]: Delete this unused field

File: EndlessClient/Dialogs/BarberDialog.cs [hunk 34]
```diff
@@ -0,0 +1,217 @@
...
+        }
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            _characterControl.Initialize();
+            UpdateListItems(_characterRepository.MainCharacter.RenderProperties);
+        }
+
+        public void Update(GameTime gameTime)
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/BarberDialog.cs:90]: Remove this unused method (this also doesn't have `override` so it could cause issues too)

File: EndlessClient/Dialogs/BarberDialog.cs (hunk 27 above)
Comment (ethanmoffat) [reply EndlessClient/Dialogs/BarberDialog.cs:77]: This doesn't appear to be resolved with the latest changes (the thick black line at the bottom of the button is incorrect). Change to `new Vector2(215, 151)`

**Review by ethanmoffat — CHANGES_REQUESTED** [review]

File: EndlessClient/Dialogs/BarberDialog.cs [hunk 35]
```diff
@@ -0,0 +1,209 @@
...
+            ApplyWidthOverride(_changeBuyHairStyleOrColor);
+
+            ApplyHighlightOffsetXOverride(_changeHairItem);
+            ApplyHighlightOffsetXOverride(_changeHairColor);
+            ApplyHighlightOffsetXOverride(_changeBuyHairStyleOrColor);
+        }
+
+        private void ApplyWidthOverride(ListDialogItem item)
+        {
+            item.WidthOverride = AdjustedWidth;
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/BarberDialog.cs:108]: I'm thinking that maybe `ListDialogItem` doesn't need to change. The `WidthOverride` value is set here, but this property isn't actually used anywhere (that I can see) except in the `SetItemSize` call below. `SetItemSize` is a wrapper around the protected method `SetSize`, which is just a convenience function for updating the `DrawArea` without changing the coordinates from within the control.

1. Back out changes to `ListDialogItem`(same as whitespace process, but for `EndlessClient/Dialogs/ListDialogItem.cs`).
2. Change the implementation of `ApplyWidthOverride` here to
```cs
item.DrawArea = new Rectangle(item.DrawArea.X, item.DrawArea.Y, AdjustedWidth, item.DrawArea.Height);
```

For style points, also do the following:
1. Mark this method as `static` (actually, double-check the other helper methods that you added and make sure they're `static` if they don't access instance members)
2. Pass in the width as a parameter to this method. This has the benefit of making the method call in `CreateListItems` a bit more readable, rather than implicitly relying on a constant defined elsewhere, it's now explicitly passed in.

File: EndlessClient/Dialogs/BarberDialog.cs [hunk 36]
```diff
@@ -0,0 +1,209 @@
...
+            item.SetItemSize(item.WidthOverride.GetValueOrDefault(232), item.Style == ListDialogItem.ListItemStyle.Large ? 36 : 13); 
+        }
+
+        private void ApplyHighlightOffsetXOverride(ListDialogItem item)
+        {
+            item.OffsetX = AdjustedHighlightXOffset;
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/BarberDialog.cs:114]: I appreciate that you're trying to make the code more readable by splitting out a bunch of helper methods. I don't know if these are really adding a lot of clarity though, especially for these one-line things. IMO it's probably better to just initialize the list items in-line so readers (me) don't have to decode the call stack and trace through what's actually happening between all these calls.

I'll leave it up to your discretion on this one (comment here with your decision so I know which direction you think is better), but I think just constructing the list items in one spot with a property initializer list would make this a lot easier to parse through.

File: EndlessClient/UIControls/CreateCharacterControl.cs [hunk 37]
```diff
@@ -68,6 +74,13 @@ public void NextHairColor()
             RenderProperties = RenderProperties.WithHairColor((RenderProperties.HairColor + 1) % 10);
         }
 
+        public void NextDirection()
```
Comment (ethanmoffat) [inline EndlessClient/UIControls/CreateCharacterControl.cs:77]: Is this method invoked anywhere? My limited "Ctrl+F" PR review search ability isn't turning up anything. If it's not used, please delete

File: EndlessClient/Dialogs/BarberDialog.cs [hunk 38]
```diff
@@ -0,0 +1,209 @@
...
+
+        private void ChangeHairStyle_Click(object sender, MonoGame.Extended.Input.InputListeners.MouseEventArgs e)
+        {
+            _characterControl.NextHairStyle();
+            _changeHairItem.SubText = GetCurrentHairStyleText(_characterControl.RenderProperties.HairStyle);
+        }
+
+        private void ChangeHairColor_Click(object sender, MonoGame.Extended.Input.InputListeners.MouseEventArgs e)
+        {
+            var currentProperties = _characterControl.RenderProperties;
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/BarberDialog.cs:170]: Unused variable here.

Side note: check out the compiler output and make sure there are no warnings.

File: EndlessClient/UIControls/BlinkingLabel.cs (hunk 19 above)
Comment (ethanmoffat) [reply EndlessClient/UIControls/BlinkingLabel.cs]: Resolved with `isUiControl` parameter when creating character renderers (to be fixed more elegantly later)

**Review by ethanmoffat — CHANGES_REQUESTED** [review]

File: EndlessClient/Dialogs/BarberDialog.cs (hunk 38 above)
Comment (ethanmoffat) [reply EndlessClient/Dialogs/BarberDialog.cs:170]: I have no idea why the compiler isn't flagging this; maybe it's because it's just a reference? Weird.

File: EndlessClient/Dialogs/BarberDialog.cs (hunk 36 above)
Comment (AstralSkies) [reply EndlessClient/Dialogs/BarberDialog.cs:114]: "every abstraction has its cost (readability in this case)". I'll go inline.

File: EndlessClient/Dialogs/BarberDialog.cs [hunk 39]
```diff
@@ -0,0 +1,200 @@
...
+        {
+            _characterControl.SetParentControl(this);
+        }
+
+        private void InitializeDialogItems(IEODialogButtonService dialogButtonService)
+        {
+            var cancel = CreateButton(dialogButtonService, new Vector2(215, 151), SmallButton.Cancel);
+            cancel.OnClick += (_, _) => Close(XNADialogResult.Cancel);
+
+            _changeHairItem = new ListDialogItem(this, ListDialogItem.ListItemStyle.Large, 0)
```
Comment (ethanmoffat) [inline EndlessClient/Dialogs/BarberDialog.cs:81]: Ok, last thing (probably):

- `OffsetY` is `25` by default for large style items. This does not need to be set here.
- Pass in `0`, `1`, and `2` as the list index for each subsequent list item (third constructor parameter, currently all zero). It automatically does the math for offsetting the list items (36*index for large style).
- Include `.DrawArea = ` and `.OffsetX = ` in the initializer lists

Comment (AstralSkies) [reply EndlessClient/Dialogs/BarberDialog.cs:81]: You will end up with the incorrect data if you add include .DrawArea in the initializer lists as there's no method of pre-knowing the math's for the draw size.

**Review by ethanmoffat — APPROVED** [review]
